Client/Frontend/Browser/SearchViewController.swift

```diff
@@ -420,6 +426,87 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
     func numberOfSectionsInTableView(_ tableView: UITableView) -> Int {
         return SearchListSection.Count
     }
+
+    func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
+        guard let section = SearchListSection.init(rawValue: indexPath.section) else {
+            return
+        }
+
+        if section == .bookmarksAndHistory,
+            let suggestion = data[indexPath.item] {
+            searchDelegate?.searchViewController(self, didHighlightText: suggestion.url, search: false)
+        }
+    }
+}
+
+extension SearchViewController {
+
+    override var canBecomeFirstResponder: Bool {
+        return true
+    }
+
+    func handleKeyCommands(sender: UIKeyCommand) {
+        let initialSection = SearchListSection.bookmarksAndHistory.rawValue
+        guard let current = tableView.indexPathForSelectedRow else {
+            if sender.input == UIKeyInputDownArrow {
+                tableView.selectRow(at: IndexPath(item: 0, section: initialSection), animated: false, scrollPosition: .top)
+            }
+            return
+        }
+
+        let currentSectionItemsCount = tableView(tableView, numberOfRowsInSection: current.section)
+        var nextSection = current.section
+        var nextItem = current.item
+        if sender.input == UIKeyInputUpArrow {
+            // we're going down, we should check if we've reached the first item in this section.
+            if (current.item == 0) {
+                // We have, so check if we can decrement the section.
+                if current.section == initialSection {
+                    // We've reached the first item in the first section.
+                    searchDelegate?.searchViewController(self, didHighlightText: searchQuery, search: true)
+                    searchDelegate?.searchViewControllerDidFinishHighlighting(self)
+                    return
+                } else {
+                    nextSection -= 1
+                    nextItem = tableView(tableView, numberOfRowsInSection: nextSection) - 1
+                }
+            } else {
+                nextItem -= 1
+            }
+        } else if sender.input == UIKeyInputDownArrow {
+            if current.item == currentSectionItemsCount - 1 {
+                if current.section == tableView.numberOfSections - 1 {
+                    return
+                } else {
+                    nextSection += 1
+                    nextItem = 0
+                }
+            } else {
+                nextItem += 1
+            }
+        }
+        let next = IndexPath(item: nextItem, section: nextSection)
+        self.tableView(tableView, didHighlightRowAt: next)
```

Comment:
Calling this delegate here feels weird. `tableView.selectRow` does the highlighting.